src/Distributed/multi_architectures.jl

```diff
@@ -0,0 +1,142 @@
+using Oceananigans.Architectures
+
+using Oceananigans.Grids: topology, validate_tupled_argument
+
+# TODO: Put connectivity inside architecture? MPI should be initialize so you can construct it in there.
+#       Might have to make it MultiCPU(; grid, ranks)
+
+abstract type AbstractMultiArchitecture <: AbstractArchitecture end
+
+struct MultiCPU{R, I, ρ, C} <: AbstractMultiArchitecture
+         my_rank :: R
+        my_index :: I
+           ranks :: ρ
+    connectivity :: C
+end
+
+child_architecture(::MultiCPU) = CPU()
+child_architecture(::CPU) = CPU()
+
+#####
+##### Converting between index and MPI rank taking k as the fast index
+#####
+
+@inline index2rank(i, j, k, Rx, Ry, Rz) = (i-1)*Ry*Rz + (j-1)*Rz + (k-1)
+
+@inline function rank2index(r, Rx, Ry, Rz)
+    i = div(r, Ry*Rz)
+    r -= i*Ry*Rz
+    j = div(r, Rz)
+    k = mod(r, Rz)
+    return i+1, j+1, k+1  # 1-based Julia
+end
+
+#####
+##### Rank connectivity graph
+#####
+
+struct RankConnectivity{E, W, N, S, T, B}
+    east :: E
+    west :: W
+   north :: N
+   south :: S
+     top :: T
+  bottom :: B
+end
+
+RankConnectivity(; east, west, north, south, top, bottom) =
+  RankConnectivity(east, west, north, south, top, bottom)
+
+function increment_index(i, R, topo)
+  R == 1 && return nothing
+  if i+1 > R
+      if topo == Periodic
+          return 1
+      else
+          return nothing
```

Comment:
Does this nothing mean that it inherets the boundary conditions imposed on the full problem?

Reply:
Hmmm yeah the function names could be improved (and comments added...).

Yes so the function is supposed to return the index of the rank to your east, north, or top to construct a `RankConnectivity`.

E.g. if there's no model to your north then the index returned is `nothing`. This is later used to inform where `HaloCommunication` boundary conditions should be injected (or not).

Reply:
Thanks for clarifying